nemo/collections/asr/modules/rnnt.py

```diff
@@ -1115,14 +1116,37 @@ def batch_replace_states_all(
         dst_states[0].copy_(src_states[0])
         dst_states[1].copy_(src_states[1])
 
+    @classmethod
     def batch_split_states(
-        self, batch_states: Tuple[torch.Tensor, torch.Tensor]
+        cls, batch_states: Tuple[torch.Tensor, torch.Tensor]
     ) -> list[Tuple[torch.Tensor, torch.Tensor]]:
         """
         Split states into a list of states.
         Useful for splitting the final state for converting results of the decoding algorithm to Hypothesis class.
         """
-        return list(zip(batch_states[0].split(1, dim=1), batch_states[1].split(1, dim=1)))
+        return [
+            (sub_state_1.squeeze(1), sub_state_2.squeeze(1))
+            for sub_state_1, sub_state_2 in zip(batch_states[0].split(1, dim=1), batch_states[1].split(1, dim=1))
+        ]
+
+    @classmethod
+    def batch_unsplit_states(
+        cls, batch_states: List[Tuple[torch.Tensor, torch.Tensor]], device=None, dtype=None
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Concatenate a batch of decoder state to a packed state.
+
+        Args:
+            batch_states (list): batch of decoder states
+                B x ([L x (H)], [L x (H)])
+
+        Returns:
+            (tuple): decoder states
+                (L x B x H, L x B x H)
+        """
+        return (
+            torch.stack([state[0] for state in batch_states], dim=1).to(device=device, dtype=dtype),
```

Comment:
In an ideal world, I think we would have a "streaming session" that holds a single contiguous memory buffer that can hold `batch_size * number_of_batches` hidden states. Then you could implement these operations via scatter kernels and gather kernels rather than splitting and concatenating. However, I don't think it is easy for NeMo to do this the way that is currently written.